File: src/main/java/com/wesker926/leetcode/algorithms/p0419/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0419.s1;

/**
 * @author wesker.gh
 * @date 2021/12/18
 * @description dfs
 */
public class Solution {
    public int countBattleships(char[][] board) {
        int count = 0;
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] != '.') {
                    count++;
                    dfs(board, i, j);
                }
            }
        }
        return count;
    }

    private void dfs(char[][] matrix, int i, int j) {
        if (i < 0 || i >= matrix.length || j < 0 || j >= matrix[0].length || matrix[i][j] == '.') {
            return;
        }
        matrix[i][j] = '.';
        dfs(matrix, i - 1, j);
        dfs(matrix, i + 1, j);
        dfs(matrix, i, j - 1);
        dfs(matrix, i, j + 1);
    }
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0419/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/18
// @description: dfs
package s1

var (
	m, n   int
	matrix [][]byte
)

func countBattleships(board [][]byte) int {
	count := 0
	matrix = board
	m, n = len(board), len(board[0])
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if matrix[i][j] != '.' {
				count++
				dfs(i, j)
			}
		}
	}
	return count
}

func dfs(i, j int) {
	if i < 0 || i >= m || j < 0 || j >= n || matrix[i][j] == '.' {
		return
	}
	matrix[i][j] = '.'
	dfs(i+1, j)
	dfs(i-1, j)
	dfs(i, j+1)
	dfs(i, j-1)
}
